core/corehttp/gateway_handler.go

```diff
@@ -133,6 +133,7 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
 	}
 
 	w.Header().Set("X-IPFS-Path", urlPath)
+	w.Header().Set("Access-Control-Allow-Origin", "*")
```

Comment:
> People will run into annoying configuration problems with it.

But will they run into problems caused by `Access-Control-Allow-Origin: *`? I don't think that they will, because I don't see a use-case for treating browsers differently than non-browser clients.

> and instead provide the user the tooling necessary to solve their own problem, whatever it might be.

But that is what tools like [nginx](http://nginx.org/en/docs/http/ngx_http_headers_module.html), [apache](http://httpd.apache.org/docs/2.2/mod/mod_headers.html), [tiny-proxy](https://banu.com/tinyproxy/), and a myriad of others are built for. It's pointless to duplicate this kind of functionality, especially when these tools are so much more powerful. They let you add headers based on conditions & locations, control access by IP &/or interface, and even allow/disallow endpoints. It's just like what you're proposing with `ipfs config Gateway.Headers...`, except it's decoupled from ipfs, it's code that doesn't have to be maintained as part of ipfs, and it's a tool built for that specific purpose, so it's naturally better at the job.

> a counter example is auth tokens and private keys

Tokens and keys are usually different for every user/instance/deployment - they need to be changed in the normal operation of the application. Headers that the HTTP server sends are not.

> facilities in a general tool dealing with the concerns of many different types of users with many different types of concerns

It shouldn't be a tool for many different types of concerns - it should only deal with ipfs :stuck_out_tongue:. Anything else should be done by another tool. To quote Doug McIlroy: "Make each program do one thing well. To do a new job, build afresh rather than complicate old programs by adding new features".

> Sure, and this applies to CORS headers too. But it's a common enough thing that it was worth considering

Yes, it does apply to CORS headers, which is why I think we should just provide a reasonable default and leave it at that.